winforms/src/toga_winforms/widgets/base.py

```diff
@@ -48,9 +32,41 @@ def update_scale(self, screen=None):
         hMonitor = windll.user32.MonitorFromRect(screen_rect, 2)
         pScale = wintypes.UINT()
         windll.shcore.GetScaleFactorForMonitor(c_void_p(hMonitor), byref(pScale))
-        Scalable._dpi_scale = pScale.value / 100
-        if not hasattr(Scalable, "original_dpi_scale"):
-            Scalable.original_dpi_scale = Scalable._dpi_scale
+        return pScale.value / 100
+
+    @property
+    def dpi_scale(self):
+        if (
+            hasattr(self, "interface")
+            and (self.interface is not None)
+            and hasattr(self.interface, "window")
+            and (self.interface.window is not None)
+        ):
+            # For Widgets and Stack Trace Dialogs
+            if issubclass(type(self), Widget) or (  # pragma: no branch
+                hasattr(self.interface.window._impl, "current_stack_trace_dialog_impl")
+                and (
+                    self.interface.window._impl.current_stack_trace_dialog_impl == self
```

Comment:
We can't assume a dialog has the same DPI as the window that created it, because the dialog could be moved to a different monitor after it appears.

Maybe dialogs and windows could be handled together by testing `isinstance(getattr(self, "native", None), WinForms.Form)`.

Reply:
Also, does this really need a `pragma: no branch`? Are there any objects which would pass the first `if` statement but not the second? If not, they should be merged into one statement.

Note that `hasattr(self, "interface") and (self.interface is not None)` can be simplified to `getattr(self, "interface", None) is not None`.

Reply:
>We can't assume a dialog has the same DPI as the window that created it, because the dialog could be moved to a different monitor after it appears.

Sorry, but I had forgot to commit the change where I had disabled the stack trace dialog to be moved to other monitor without moving its window. This is because, I had found that on other platforms, the dialogs cannot be moved separately without moving the window and there was also some other reasons, but I can't remember it. As for other dialogs, windows scales them up automatically when they move from one monitor to another with different DPI scale.

>Also, does this really need a pragma: no branch? Are there any objects which would pass the first if statement but not the second? If not, they should be merged into one statement.

Thanks. I have removed the second statement as it was not really required.

Reply:
> I had disabled the stack trace dialog to be moved to another monitor without moving its window. This is because, I had found that on other platforms, the dialogs cannot be moved separately without moving the window and there were also some other reasons, but I can't remember it.

Other platforms have other conventions, but on Windows you can normally move dialogs separately from their windows, so unless you can remember the other reasons, you'd better revert this. The current behavior of the dialog fighting with the user to get back to its original position is weird and glitchy.

Also, this dialog isn't resizing itself along with its content:
* If you open it at 100%, and then switch to 125%, its content no longer fits in its borders, so you can't even click the OK button to close it. 
* If you open it at 125% and then switch to 100%, it now has a big gap on the right and bottom edges. And whether the font sizes scale correctly appears to depend on whether the scale was 100 or 125% when the app was *started*.

Since dialogs are ephemeral, if there's a way to avoid all of the complexity around the stack trace dialog by falling back on the default blurry scaling, I think that would be acceptable. It looks like that's what's happening with the simple question/info/etc. dialogs anyway.

Reply:
The dialog isn't properly scaling because, I have removed this branch line: [line](https://github.com/beeware/toga/pull/2155#:~:text=if%20isinstance(control%2C%20WinForms.Button)%3A%20%20%23%20or%20isinstance( )

>Other platforms have other conventions, but on Windows you can normally move dialogs separately from their windows, so unless you can remember the other reasons, you'd better revert this. The current behavior of the dialog fighting with the user to get back to its original position is weird and glitchy.
...
Since dialogs are ephemeral, if there's a way to avoid all of the complexity around the stack trace dialog by falling back on the default blurry scaling, I think that would be acceptable. It looks like that's what's happening with the simple question/info/etc. dialogs anyway.

In that case I'll remove the scaling from dialogs and push a commit.